src/test/kotlin/org/rust/ide/typing/assist/RsSmartEnterProcessorTest.kt

```diff
@@ -300,6 +300,54 @@ class RsSmartEnterProcessorTest : RsTestBase() {
         }
     """)
 
+    fun `test fix function`() = doTest("""
+        fn foo/*caret*/
+    """, """
+        fn foo() {
+        /*caret*/
+        }
+    """)
+
+    fun `test fix function with parameters`() = doTest("""
+        fn foo(a: i32, b: i32/*caret*/)
+    """, """
+        fn foo(a: i32, b: i32) {
+        /*caret*/
+        }
+    """)
+
+    fun `test fix function with return type`() = doTest("""
+        fn foo() -> i32/*caret*/
+    """, """
+        fn foo() -> i32 {
+        /*caret*/
+        }
+    """)
+
+    fun `test fix function with parameters and return type`() = doTest("""
+        fn foo(a: i32) -> i32/*caret*/
+    """, """
+        fn foo(a: i32) -> i32 {
+        /*caret*/
+        }
+    """)
+
+    fun `test fix struct`() = doTest("""
+        struct s/*caret*/
+    """, """
+        struct s {
+        /*caret*/
+        }
+    """)
+
+    fun `test fix union`() = doTest("""
+        union u/*caret*/
+    """, """
+        union u {
+        /*caret*/
+        }
+    """)
```

Comment:
Some other cases:

```rust
// Before
fun foo(/*caret*/)

// After
fn foo() {
    /*caret*/
}
```

```rust
// Before
fn foo(/*caret*/

// After
fn foo() {
    /*caret*/
}
```

```rust
// Before
fn foo() {
    fn bar(/*caret*/)
}

// After
fn foo() {
    fn bar() {
        /*caret*/
    }
}
```

Reply:
Also, you need to reformat function/struct body. Kotlin does it in `KotlinSmartEnterHandler.moveCaretInsideBracesIfAny`:
https://github.com/JetBrains/kotlin/blob/master/idea/src/org/jetbrains/kotlin/idea/editor/KotlinSmartEnterHandler.kt#L111

Reply:
> Also, you need to reformat function/struct body. Kotlin does it in `KotlinSmartEnterHandler.moveCaretInsideBracesIfAny`:
> https://github.com/JetBrains/kotlin/blob/master/idea/src/org/jetbrains/kotlin/idea/editor/KotlinSmartEnterHandler.kt#L111

After calling `reformat`, it will become like this.

```rust
// Before
fn foo(/*caret*/)

// After
fn foo() {/*caret*/}
```

Reply:
@mchernyavsky As far as the current parsing results are concerned, I suggest not to implement `fn foo(/*caret*/` `Complete the current statement`.

For example, the following expressions will generate some "inconvenient to parse" tree.
```rust
fn main() {
    fn bar(/*caret*/

    println!("Hello, world!");
}
```

```rust
fn foo(/*caret*/

fn main() {
    println!("Hello, world!");
}
```

```rust
fn foo() {
    fn bar(/*caret*/

    return;
}
```

Reply:
ping @mchernyavsky